src/Products/ZCatalog/Catalog.py

```diff
@@ -897,29 +886,83 @@ def _sort_nbest_reverse(self, actual_result_count, result, rs,
                         n += 1
                     best = keys[-1]
         else:
-            for did in rs:
+            # we have multi index sorting
+            result = self._multi_index_nbest(
+                actual_result_count,
+                result,
+                rs,
+                limit,
+                sort_index,
+                sort_spec,
+                second_indexes_key_map,
+                reverse=False
+            )
+
+        return (actual_result_count, 0, result)
+
+    def _multi_index_nbest(self, actual_result_count, result,
+                           rs, limit, sort_index, sort_spec,
+                           second_indexes_key_map, reverse=True):
+        """
+        For multiple indexes.
+        1) Categorize documents as lists by the first index values in the
+        did_by_index_value dict.
+        2) Sort the index_values
+        3) Collect from the did_by_index_value dict by the sorted
+        index_values till limit is exeeded.
+        """
+        # A dict of lists categorize the documents after the first sort
+        # index value.
+        did_by_index_value = {}
+        # get the index' keymap
+        index_key_map = sort_index.documentToKeyMap()
+        # for all documents
+        for did in rs:
+            # get the index value of the current document id
+            try:
+                index_value = index_key_map[did]
+            except KeyError:
+                # This document is not in the sort key index, skip it.
+                # ToDo: Is this the correct/intended behavior???
+                actual_result_count -= 1
+            else:
+                # do we already have a list for this index_value? If not
+                # create one.
+                if index_value not in did_by_index_value:
+                    did_by_index_value[index_value] = []
+                # add document id for the index value
+                did_by_index_value[index_value].append(did)
+        # All documents are now categorized after the first sort index values.
+        # Sort the sort index_values
+        sorted_index_values = sorted(
+            did_by_index_value.keys(),
+            reverse=reverse)
```

Comment:
While this is correct, you lose the former optimization (use `nbest` rather than full sorting to save sorting time). This may be relevant if the first sort index is large (e.g. a timestamp based index such as `modified`, `effective`, ...). I would avoid the full sorting and use `heapq.n[smallest|largest]` (depending on `reverse`) with *limit*\ ` + 1` as `*n`. You might even use more elementary `heapq` functions to sort incrementally and stop as soon as you have enough documents found.

Reply:
I do not think that we can avoid a full sort (on the first index). Coming back to my example from above:

Sort_on col 1, col2 : n_start = 0, n_size=2 -> limit = 2
Data:
eggs 1
ham 2
ham 1
spam 2
... lots of elements with no eggs!...
ham 0

Correct result:
eggs 1
ham 0

In this case we need to sort all of the first index to find all element with "ham" and crucially "ham 0". It does not help that we find with an incremental sort the elements 'ham 1" and "ham 2" which fulfills the criterion limit=2. Without finding also "ham 0" the sort on the second index fails to deliver the correct result.

Cheers,
Volker

Reply:
After digging into heapq: Heapq asumes a fixed size heap.
So if say limit=2 but I need in reality three elements
Result-set:
eggs 1
ham 2
ham 1

Expected Result set:
eggs 1
ham 1

I would start with a heap of length limit=2. Then I would have to extend the heap iteratevely and push in all the index values in again.

I thinkt that it may be preferable to switch to a linear search here. Finding all results with the same index value of the largest/smallest value (not already in the sort body) is of O(n) and this operation takes place only once.

Cheers,
Volker

Reply:
Sorry for the delay. Will work on this at the weekend

Reply:
Finally! 

I pushed a new version using heapq.

> Ideally, sort_nbest and sort_nbest_reverse would be symmetric. Your change above seems to have reduced the symmetry.
> 
> You have refactored the multi-index sorting into a single method. I would go a step further and merge sort_nbest and sort_nbest_reverse into a single function (in my view those methods are also named unintuitively; the heapq module contains similar functions named nsmallest and nlargest, respectively - which would be much better names). I would use functools.partial to instantiate the proper sorting function from it.

Currently these methods utilize a bisecting scheme that is probably as effective as heapq, if the bisecting is implemented in C-Code. One can replace this code by the using heapq to use the same sorting algorithm.

I start doing this now

Volker

Reply:
My heapq Code is currently sorting the index_values [primitive], only.
If we like to handle the former bisec_code we have to sort a list of tuples [(primitive, did)].

A benchmark for 100Mio values to sort [primitive] or [(primitive, did)] I got

import heapq
import random
import datetime

NUM = 100000000

a = []
b =[]

for i in range(NUM):

    a.append(random.random())
    b.append((random.random(), random.random()))

start = datetime.datetime.now()

x = heapq.nlargest(100, a)

now = datetime.datetime.now()
diff = now - start
print( diff.total_seconds())

start = datetime.datetime.now()

y = heapq.nlargest(100, b)

now = datetime.datetime.now()
diff = now - start

print( diff.total_seconds())




8.796026
11.152541

One will need  27% more time for heapq-sort to sort a list of tuples than only primitives.

So I am uncertain which way to go. 

Cheers,
Volker